src/session.c

```diff
@@ -232,20 +435,93 @@ CK_RV p11prov_get_session(P11PROV_CTX *provctx, CK_SLOT_ID *slotid,
         *next_slotid = CK_UNAVAILABLE_INFORMATION;
     }
 
-    p11prov_ctx_unlock_slots(provctx, &slots);
-
     if (ret != CKR_OK) {
         return ret;
     }
 
-    sess = p11prov_session_new(provctx, id);
-    if (sess == NULL) {
-        return CKR_HOST_MEMORY;
+    ret = CKR_OK;
+
+    /* LOCKED SECTION ------------- */
+    pthread_mutex_lock(&slot->pool->lock);
+    for (i = 0; i < slot->pool->num_sessions; i++) {
+        if (slot->pool->sessions[i]->free) {
+            bool ok;
+            int expected = 1;
+
+            session = slot->pool->sessions[i];
+
+            ok =
+                __atomic_compare_exchange_n(&session->free, &expected, 0, false,
+                                            __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
+            if (!ok) {
+                P11PROV_raise(provctx, CKR_GENERAL_ERROR,
+                              "Unexpectd busy session while holding lock");
+                ret = CKR_GENERAL_ERROR;
+            }
+            break;
+        }
     }
-    ret = p11prov_session_open(sess, false, NULL, 0);
+    pthread_mutex_unlock(&slot->pool->lock);
+    /* ------------- LOCKED SECTION */
+
     if (ret != CKR_OK) {
         return ret;
     }
-    *session = sess;
-    return CKR_OK;
+
+    if (session == NULL) {
+        session = p11prov_session_new(provctx, slot);
+        if (session == NULL) {
+            return CKR_GENERAL_ERROR;
+        }
+    }
+
+    /* ref now, so we can simply p11prov_ession_free() later on errors */
+    session = p11prov_session_ref(session);
+    if (session == NULL) {
+        P11PROV_raise(provctx, CKR_GENERAL_ERROR,
+                      "Failed to ref count session");
+        /* intentionally leave this broken session busy so it won't be
+         * used anymore */
+        return CKR_GENERAL_ERROR;
+    }
+
+    if (session->session != CK_INVALID_HANDLE) {
+        /* check that the pkcs11 session is still ok */
+        CK_FUNCTION_LIST *f;
+        CK_SESSION_INFO session_info;
+
+        f = p11prov_ctx_fns(provctx);
+        if (f == NULL) {
+            ret = CKR_GENERAL_ERROR;
```

Comment:
Most other code paths use `CKR_FUNCTION_FAILED` instead. But consider making it `p11prov_ctx_fns` contract  to return valid pointer or abort.

Reply:
See the other comment, I want to try to make the provider stop working entirely and return a CKR_GENERAL_ERROR everywhere, in fact I think I will change the prototype of the function to return an error and get a pointer to fns as an argument to fill in instead.

Reply:
See #53